src/Storage/QueryBuilder.php

```diff
@@ -0,0 +1,169 @@
+use Psr\Log\LoggerInterface;
+
+class QueryBuilder {
+
+  protected $tables = [];
+
+  protected $columns = [];
+
+  public function __construct(
+    protected FieldStorageDefinitionInterface $fieldStorageDefinition,
+    protected EntityTypeManagerInterface $entityTypeManager,
+    protected Connection $database,
+    protected LoggerInterface $logger,
+  ) {
+    $tableMapping = $this->entityTypeManager->getStorage($this->fieldStorageDefinition->getTargetEntityTypeId())->getTableMapping();
+
+    // Get table definitions.
+    $table_names = $tableMapping->getAllFieldTableNames($this->fieldStorageDefinition->getName());
+    $this->tables = [
+      'entity' => $table_names[0],
+    ];
+    $this->tables['entity_revision'] = !empty($table_names[1]) ? $table_names[1] : $table_names[0];
+
+    // Get column definitions.
+    $base_field_id = $this->fieldStorageDefinition->isBaseField() ? 'id' : 'entity_id';
+    $this->columns = [
+      'id' => $base_field_id,
+      'revision_id' => $this->fieldStorageDefinition->isRevisionable() ? 'revision_id' : $base_field_id,
+      'target_id' => $tableMapping->getFieldColumnName($this->fieldStorageDefinition, 'target_id'),
+      'weight' => $tableMapping->getFieldColumnName($this->fieldStorageDefinition, 'weight'),
+    ];
+  }
+
+  private function getTablePrefix() {
+    return $this->database->tablePrefix();
+  }
+
+  protected function getAncestorSql(): string {
+    $table_name = $this->getTablePrefix() . $this->tables['entity'];
+    $revision_table_name = $this->getTablePrefix() . $this->tables['entity_revision'];
+    $column_id = $this->columns['id'];
+    $column_revision_id = $this->columns['revision_id'];
+    $column_target_id = $this->columns['target_id'];
+    $column_weight = $this->columns['weight'];
+    $sql = <<<CTESQL
+WITH RECURSIVE ancestors AS
+(
+  SELECT $column_id AS id, $column_revision_id as revision_id, $column_target_id AS target_id, $column_weight AS weight, 0 AS depth FROM $revision_table_name WHERE $column_id = :id AND $column_revision_id = :revision_id
+  UNION ALL
+  SELECT c.$column_id, c.$column_revision_id, c.$column_target_id, c.$column_weight, ancestors.depth-1 FROM $table_name c
+  JOIN ancestors ON c.$column_id=ancestors.target_id
+)
+CTESQL;
+    return $sql;
+  }
+
+  /**
+   * @return \Drupal\entity_hierarchy\Storage\Record[]
+   */
+  public function findAncestors(ContentEntityInterface $entity): RecordCollection {
+    $sql = $this->getAncestorSql() . " SELECT * FROM ancestors ORDER BY depth";
+    $result = $this->database->query($sql, [
+      ':id' => $entity->id(),
+      ':revision_id' => $entity->getRevisionId() ?: $entity->id(),
+    ]);
+    $records = $result->fetchAll(\PDO::FETCH_CLASS, '\Drupal\entity_hierarchy\Storage\Record');
+    $type = $this->fieldStorageDefinition->getTargetEntityTypeId();
+    array_walk($records, function (Record $record) use ($type) {
+      $record->setType($type);
+    });
```

Comment:
I wonder if we could use the third arg to fetchAll to pass the type as a constructor arg and avoid this?